### ISSUE TITLE
How to handle hotfix releases to production?

### ISSUE THREAD

**Issue by brandoncollins7** [saenyakorn/monorepo-versioning-gitops#102]
Let's say I have published my apps to production and have begun working on the next release on beta. Now I have realized there is a minor bug on production but all my changes on staging have not been properly tested yet and are too risky to push to production. So I would like to be able to create a branch from the last release on main fix the super minor bug and deploy that change to production?

I am trying to understand how the versioning would work given that production might be on version 1.0.0 and beta may be on 1.0.1-beta.0. Simply versioning and deploying a hotfix to production would make it 1.0.1 and somehow beta then needs to get bumped to 1.0.2-beta.0?

Or is there a way I can version production hotfix to 1.0.0-hotifx.0? But the problem here is creating any changeset for the hotfix release will cause it to become 1.0.1-hotfix.0 if I try to use a hot fix prerelease....

Any ideas on how to handle this scenario?

**Comment by saenyakorn** [comment]
Thanks @brandoncollins7  for raising the interesting topic. Assume that the hotfix you would like to push to the main branch is super super minor bug that would not conflict the beta (staging) branch. You might have to branch `hotfix/something`  directly from the `main` branch (instead of branching from `beta` in the normal flow).   

If you're not sure that your hotfix would fix the bug, you have to merge `hotfix/something` into `dev` branch first to let other reviewers to test the logic before open the PR to the main. (or you can skip this step)

After you're open the hotfix PR to the `main` branch, you have to add Changeset file to describe the hotfix change as a `patch`. Let say that the current version of your app is `web@1.0.2`, after you merged the PR, the new version should be bumped to `web@1.0.3`. Meanwhile, the beta (staging) does not contain this changes. You need to rebase `beta` to `main` branch after the hotfix was released. All of the Changeset file in beta will remain the same. 

To summarize, the steps you should take are

1. Branching from `main` as `hotfix/something`
2. Fix the bug in `hotfix/something`
3. (Optional) Merge your `hotfix/something` into the dev branch to test the logic before go live
4. Open the PR from `hotfix/something` to `main` branch 
5. Also, don't forget to add Changeset file to describe your hotfix change
6. After you got all approved, merge it. The "Version Packages" will be appear. Then merge it. The version should be bump from `web@1.0.2` to `web@1.0.3` (For example)
7. Then, rebase the `beta` to `main` to add your hoxfix logic to beta branch. 
8. All done!

Thank to @brandoncollins7  again. I totally forgot to add this topic in the article. I will add it later.

Do my response answer your question?

**Comment by brandoncollins7** [comment]
But what if you release 1.0.3 as a hotfix but in the meantime beta has already released 1.0.3-beta.1? The automated pr's get confused since now beta should actually be 1.0.4-beta.0?

**Comment by saenyakorn** [comment]
TL;DR please rebase beta onto main branch after the hotfix was merged into main and the packages were bumped their versions. The PR description of "Version Packages (Beta)" would be update automatically after rebasing. Additionally, the "Version Packages"'s PR does not bump the version but it describes how version changes "IF" we merged the PR.

Here's my experiment on Changesets CLI for the hotfix case.

1. Created a new Monorepo and setup Changesets by running `npx changeset init` (Let says that I have `web` and `docs` as apps and `ui` as packages with version 0.0.0)
3. Created a beta branch by `git switch -C beta`
4. Configured the Changesets config to enter pre-release mode by `npx changeset pre enter beta`
5. Bumped `docs` as patch; the description was "[BETA] just for testing"
6. Committed changes to beta
7. Switched back to the main and mock the hotfix commit of `docs` as patch; the description was "[Main] Mock hotfix" and version it by `npx changeset version` (The docs' version should be bumped to `0.0.1`)
9. Commit the changes in main branch
10. After the 7th step, the hotfix should be merged and bumped version in the main branch. Now, try to switch back to the beta branch and rebase beta onto main.
11. Try to run `npx changeset version`, you may see that generated `CHANGELOG` indicated the changes as `0.0.2-beta.0` , not `0.0.1-beta.0`

PS. It works like how PR's description is generated 

Here's the result by following the step above

<img width="1680" alt="Screenshot 2566-05-07 at 12 01 01" src="https://user-images.githubusercontent.com/33742791/236658615-39a74b60-9628-48b8-8505-f0dfb83a76a6.png">

**Comment by brandoncollins7** [comment]
Thanks very much for taking the time and coming up with this process. I've been trying to get this all just right for a long time:)

So when I tried this, the sync beta to main workflow ran and failed. Is this expected in your flow? Confused because you are saying to manually rebase but this workflow "should" already do it automatically when pr is closed to main. But I did get this error? Is expected? 
![image](https://user-images.githubusercontent.com/36413482/236852849-8f379fb4-69ed-4e77-8459-06d64de2472a.png)

**Comment by saenyakorn** [comment]
Yeah, you're right. The workflow "should" do the rebase task properly. I think I would retry my experiment again in the real repository (Not mocking Changeset with its CLI like above). I'll let you know if I finished. Or, if you have solve the problem, you can share it to me :) 

Thanks for raising the problem

**Comment by brandoncollins7** [comment]
Let me know what you think of this:) https://github.com/saenyakorn/monorepo-versioning-gitops/pull/115

**Comment by saenyakorn** [comment]
Closed #115
